ouroboros-network/src/Ouroboros/Network/Connections/Concurrent.hs

```diff
@@ -6,138 +6,215 @@
+  -> State identifier handle
+insertConnection identifier conn state = state
+  { connectionMap    = Map.insert identifier numConn (connectionMap state)
+  , connectionNumber = connectionNumber state + 1
+  }
+  where
+  !numConn = NumberedConnection
+    { number     = connectionNumber state
+    , connection = conn
+    }
+
+removeConnection
+  :: ( Ord identifier )
+  => identifier
+  -> State identifier handle
+  -> State identifier handle
+removeConnection identifier state =
+  let alteration mentry = case mentry of
+        -- removeConnection is called by the thread that was spawned by the
+        -- handler before insertConnection happened for this identifier. If
+        -- there's nothing here then there's a bug in this module.
+        Nothing -> error "bug: entry vanished"
+        Just _  -> Nothing
+  in  state { connectionMap = Map.alter alteration identifier (connectionMap state) }
+
+-- | Generic concurrent connection manager in which at most one connection for
+-- a given identifier is allowed. New connections will re-use existing
+-- connections, and existing connections will be rejected if there is already
+-- one here.
+--
+-- For example, we could choose
+--   resource := Socket
+--   connectionId := exists sockType . (SockAddr sockType, SockAddr sockType)
+-- and this would guarantee at most one connection per socket address pair,
+-- when the sockType is IPv* and it's a TCP socket.
+-- However, for Unix domain sockets, this is no good, because unnamed sockets
+-- all have (==) SockAddr values. A richer connection identifier type would be
+-- needed, and any servers or clients calling into this concurrent connection
+-- manager with Unix domain sockets would be required to come up with unique
+-- identifiers for unnamed ones (probably by taking the file descriptor number).
 concurrent
-  :: (Socket -> SockAddr -> IO ())
-  -> (Connections SockAddr Socket Reject Accept IO -> IO t)
+  :: forall connectionId resource handle reason t .
+     ( Ord connectionId )
+  => (connectionId -> resource -> IO (Either reason (Handler handle)))
+  -- ^ A callback to run for each connection. The `handle` gives an interface
+  -- to that connection, allowing for inspection and control etc. of whatever
+  -- it's doing.
+  -- If this throws an exception, it is re-thrown, so that calls to
+  -- `include` will get them.
+  -- Non-exceptional domain-specific reasons to reject a connection are given
+  -- in the `Left` variant.
+  -> (Connections connectionId resource (Reject reason) (Accept handle) IO -> IO t)
   -> IO t
-concurrent withSocket k = do
-  -- Cannot use an IORef for state because spawned threads will modify this
-  -- when they finish.
-  state :: MVar (Map SockAddr Connection) <- newMVar Map.empty
-  let connections = Connections { include = includeOne state }
-  k connections `finally` killThreads state
+concurrent withResource k = do
+  stateVar :: MVar (State connectionId handle) <- newMVar $ State
```

Comment:
Why is this an MVar and not a StrictTVar?

Reply:
How will the connection tables be shared between the governor and the server? We need to run one `concurrent` for IPv4 and one for IPv6 for the server (same port number but different socket to listen to), and at least one `concurrent` for the client side.
Technically we can have a different table for IPv4 and IPv6 since the connections can never be shared, but the client still must be able to access both tables since it needs to be able to talk to both types of peers.

Reply:
It's an `MVar` because that's IMO the most appropriate, simplest thing to use (i.e. anyone else writing this who doesn't know about io-sim would have chosen it). I'll have to change it to be a `TMVar` in order to make this thing fit with io-sim. Is there a `StrictTMVar` too?

Reply:
> How will the connection tables be shared between the governor and the server?

This `concurrent` function takes care of it. So long as the governor and the server use the same `Connections` term, it's all good. 

> We need to run one concurrent for IPv4 and one for IPv6 for the server (same port number but different socket to listen to), and at least one concurrent for the client side.

In fact, we'll have just _one_ for all node-to-node connections. We can run two different accept loops using the same manager, one for IPv6 and one for IPv4 (could even use UNIX sockets as well but they have identifier problems).

For client side, you can give any bind and connect address and use that same `Connections` term to try and bring up a connection.

> Technically we can have a different table for IPv4 and IPv6 since the connections can never be shared, but the client still must be able to access both tables since it needs to be able to talk to both types of peers.

The table will be the same.

Reply:
I think @karknu question makes sense: you don't need to await when the shared stated is created, so between `TVar` and `TMVar`, `TVar` makes more sense. 

Some bangs can be removed when using `MonadSTM.Strict`.

Reply:
I'll switch over to `MonadSTM.Strict`, but it really must be `MVar`-style semantics, because we need to do IO while holding the lock.